sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala

```diff
@@ -88,7 +88,10 @@ class StringType private(val collationId: Int) extends AtomicType with Serializa
  */
 @Stable
 case object StringType extends StringType(0) {
-  private[spark] def apply(collationId: Int): StringType = new StringType(collationId)
+  private[spark] def apply(collationId: Int): StringType = {
+    assert (collationId >= 0 && collationId <= (1 << 12))
```

Comment:
where did we check collation id before?

Reply:
It did not check `collationId`, it only checked `collationName`,
https://github.com/apache/spark/blob/1a428c1606645057ef94ac8a6cadbb947b9208a6/sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala#L94
https://github.com/apache/spark/blob/1a428c1606645057ef94ac8a6cadbb947b9208a6/common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java#L391
https://github.com/apache/spark/blob/1a428c1606645057ef94ac8a6cadbb947b9208a6/common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java#L615

Reply:
Additionally, exceptions may be thrown below due to invalid `collationId`, but unfortunately, they are all `functions `that only `trigger` when `called`
https://github.com/apache/spark/blob/83a97917d46debe9c3ebb24cf2540b07f6a902ec/sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala#L32-L46